<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Probabilistic Factor Analysis for Spatially-Aware Dimension Reduction • ProFAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Probabilistic Factor Analysis for Spatially-Aware Dimension Reduction">
<meta property="og:description" content="Probabilistic factor analysis for spatially-aware dimension reduction across multi-section spatial transcriptomics data with millions of spatial locations.
    More details can be referred to Wei Liu, et al. (2023) &lt;doi:10.1101/2023.07.11.548486&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ProFAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/FASTdlpfc2.html">FAST: two DLPFC sections</a>
    </li>
    <li>
      <a href="articles/FASTsimu.html">FAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/ProFAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="profast">ProFAST<a class="anchor" aria-label="anchor" href="#profast"></a>
</h1></div>
<p>========================================================================= <!-- badges: start --></p>
<p><a href="https://cran.r-project.org/package=ProFAST" class="external-link"><img src="https://www.r-pkg.org/badges/version-ago/ProFAST"></a> <a href="https://cran.r-project.org/package=ProFAST" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/ProFAST?color=orange"></a> <a href="https://cran.r-project.org/package=ProFAST" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/ProFAST?color=orange"></a> <!-- badges: end --></p>
<p>FAST is an advanced probabilistic factor analysis technique designed for spatially-aware dimension reduction in multi-section spatial transcriptomics data. The ‘ProFAST’ package incorporates the FAST method and is specifically developed by the Jin Liu’s lab for the comprehensive analysis of multiple spatially resolved transcriptomics (SRT) datasets.</p>
<p>Check out our <a href="https://feiyoung.github.io/ProFAST/index.html" class="external-link">Package Website</a> for a more complete description of the methods and analyses.</p>
<p>ProFAST can be used to compare and contrast experimental datasets in a variety of contexts, for instance:</p>
<ul>
<li>Across experimental batches</li>
<li>Across individuals</li>
<li>Across different conditions (i.e., case and control)</li>
<li>Across datasets with only partially shared cell/domain clusters</li>
</ul>
<p>Once the embeddings of multiple datasets are estimated by ProFAST, the package provides functionality for further data exploration, analysis, and visualization. Users can:</p>
<ul>
<li>Align the embeddings obtained by ProFAST</li>
<li>Identify clusters using all data information</li>
<li>Recover comparable gene expression matrices among datasets</li>
<li>Find significant shared (and dataset-specific) gene markers</li>
<li>Visuzlize extracted embeddings using 3-dim tSNE and UMAP</li>
<li>Visualize clusters and gene expression using tSNE and UMAP</li>
</ul>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>“ProFAST” depends on the ‘Rcpp’ and ‘RcppArmadillo’ package, which requires appropriate setup of computer. For the users that have set up system properly for compiling C++ files, the following installation command will work.</p>
<pre class="{rmd}"><code><span class="co"># Method 1: install ProFAST from CRAN</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'ProFAST'</span><span class="op">)</span>


<span class="co"># For the newest version of ProFAST, users can use method 2 for installation.</span>
<span class="co"># Method 2: Install ProFAST from Github</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"feiyoung/ProFAST"</span><span class="op">)</span>

<span class="co"># If some dependent packages (such as `scater`) on Bioconductor can not be installed nomrally, use following commands, then run abouve command.</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="co">## install BiocManager</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="co"># install the package on Bioconducter</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span><span class="op">)</span></code></pre>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For usage examples and guided walkthroughs, check the <code>vignettes</code> directory of the repo.</p>
<ul>
<li><a href="https://feiyoung.github.io/ProFAST/articles/FASTsimu.html" class="external-link">Integration analysis for a toy example with three batches</a></li>
<li><a href="https://feiyoung.github.io/ProFAST/articles/FASTdlpfc2.html" class="external-link">Integration analysis for DLPFC data with two batches</a></li>
</ul>
<p>For the users that don’t have set up system properly, the following setup on different systems can be referred. ## Setup on Windows system First, download <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a>; second, add the Rtools directory to the environment variable.</p>
</div>
<div class="section level2">
<h2 id="setup-on-macos-system">Setup on MacOS system<a class="anchor" aria-label="anchor" href="#setup-on-macos-system"></a>
</h2>
<p>First, install Xcode. Installation about Xcode can be referred <a href="https://stackoverflow.com/questions/8291146/xcode-installation-on-mac" class="external-link">here</a>.</p>
<p>Second, install “gfortran” for compiling C++ and Fortran at <a href="https://github.com/fxcoudert/gfortran-for-macOS" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="setup-on-linux-system">Setup on Linux system<a class="anchor" aria-label="anchor" href="#setup-on-linux-system"></a>
</h2>
<p>If you use conda environment on Linux system and some dependent packages (such as <code>scater</code>) can not normally installed, you can search R package at anaconda.org website. We take the <code>scater</code> package as example, and its search result is <a href="https://anaconda.org/bioconda/bioconductor-scater" class="external-link uri">https://anaconda.org/bioconda/bioconductor-scater</a>. Then you can install it in conda environment by following command.</p>
<pre class="{linux}"><code>
conda install -c bioconda bioconductor-scater</code></pre>
<p>For the user not using conda environment, if dependent packages (such as <code>scater</code>) not normally installed are in Bioconductor, then use the following command to install the dependent packages.</p>
<pre class="{linux}"><code><span class="co"># install BiocManager</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="co"># install the package on Bioconducter</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span><span class="op">)</span></code></pre>
<p>If dependent packages (such as <code>DR.SC</code>) not normally installed are in CRAN, then use the following command to install the dependent packages.</p>
<pre class="{linux}"><code><span class="co"># install the package on CRAN</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"DR.SC"</span><span class="op">)</span></code></pre>
</div>
</div>
<div class="section level1">
<h1 id="demonstration">Demonstration<a class="anchor" aria-label="anchor" href="#demonstration"></a>
</h1>
<p>For an example of typical ProFAST usage, please see our <a href="https://feiyoung.github.io/ProFAST/index.html" class="external-link">Package Website</a> for a demonstration and overview of the functions included in ProFAST.</p>
</div>
<div class="section level1">
<h1 id="news">NEWs<a class="anchor" aria-label="anchor" href="#news"></a>
</h1>
<ul>
<li><p>ProFAST version 1.3 (2024-01-09)</p></li>
<li><p>ProFAST version 1.1 (2023-06-03)</p></li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=ProFAST" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/feiyoung/ProFAST/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/feiyoung/ProFAST/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ProFAST</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Wei Liu <br><small class="roles"> Author, maintainer </small>  </li>
<li>Xiao Zhang <br><small class="roles"> Author </small>  </li>
<li>Jin Liu <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
