<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CoFAST: NSCLC CosMx data coembedding • ProFAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CoFAST: NSCLC CosMx data coembedding">
<meta property="og:description" content="ProFAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ProFAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CosMx.html">CoFAST: NSCLC CosMx data coembedding</a>
    </li>
    <li>
      <a href="../articles/FASTdlpfc.html">FAST: single DLPFC section</a>
    </li>
    <li>
      <a href="../articles/FASTdlpfc2.html">FAST: two DLPFC sections</a>
    </li>
    <li>
      <a href="../articles/FASTsimu.html">FAST: simulation</a>
    </li>
    <li>
      <a href="../articles/pbmc3k.html">CoFAST: PBMC scRNA-seq data coembedding</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/ProFAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CoFAST: NSCLC CosMx data coembedding</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2024-03-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/ProFAST/blob/HEAD/vignettes/CosMx.Rmd" class="external-link"><code>vignettes/CosMx.Rmd</code></a></small>
      <div class="hidden name"><code>CosMx.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the CoFAST workflow for the analysis of NSCLC CosMx spatial transcriptomics dataset. In this vignette, the workflow of CoFAST consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Coembedding dimension reduction</li>
<li>Downstream analysis (i.e. , signature gene analysis, visualization of cell types and coembeddings,)</li>
</ul>
<div class="section level2">
<h2 id="load-and-view-data">Load and view data<a class="anchor" aria-label="anchor" href="#load-and-view-data"></a>
</h2>
<p>We demonstrate the use of CoFAST to NSCLC data, which can be downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span> <span class="co"># set a random seed for reproducibility.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/ProFAST" class="external-link">ProFAST</a></span><span class="op">)</span> <span class="co"># load the package of FAST method</span>
<span class="co">#&gt; Loading required package: gtools</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span>
<span class="va">CosMx_subset</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 960 features across 2814 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (960 features, 960 variable features)</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'Seurat' was built under R version 4.1.3</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; Attaching sp</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>First, we normalize the data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></code></pre></div>
<p>Then, we select the variable genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="coembedding-using-fast">Coembedding using FAST<a class="anchor" aria-label="anchor" href="#coembedding-using-fast"></a>
</h2>
<p>We introduce how to use FAST to perform coembedding for this CosMx data. First, we determine the dimension of coembeddings. Then, we select the variable genes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diagnostic.cor.eigs.html">diagnostic.cor.eigs</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">q_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">dat_cor</span>, <span class="st">"q_est"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"q_est = "</span>, <span class="va">q_est</span>, <span class="st">'\n'</span><span class="op">)</span>
<span class="co">#&gt; q_est =  28</span></code></pre></div>
<p>Subsequently, we calculate coembeddings by utilizing FAST, and observe that the <code>reductions</code> field acquires an additional component named <code>fast</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co"># Extract the spatial coordinates</span>
<span class="va">Adj_sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span> <span class="co">## calculate the adjacency matrix</span>
<span class="co">#&gt; The spatial cooridnates are 2 dimensions</span>
<span class="co">#&gt; Find the adjacency matrix by bisection method...</span>
<span class="co">#&gt; Current radius is 62.34</span>
<span class="co">#&gt; Median of neighborhoods is 0</span>
<span class="co">#&gt; Current radius is 93.01</span>
<span class="co">#&gt; Median of neighborhoods is 2</span>
<span class="co">#&gt; Current radius is 93.01</span>
<span class="co">#&gt; Median of neighborhoods is 5</span>
<span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NCFM_fast.html">NCFM_fast</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, Adj_sp <span class="op">=</span> <span class="va">Adj_sp</span>, q <span class="op">=</span> <span class="va">q_est</span><span class="op">)</span>
<span class="co">#&gt; Finish variable intialization </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 2, loglik= -1426036.505262, dloglik=0.999336 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 3, loglik= -1412722.725309, dloglik=0.009336 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 4, loglik= -1409935.707519, dloglik=0.001973 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 5, loglik= -1408112.785855, dloglik=0.001293 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 6, loglik= -1407078.327983, dloglik=0.000735 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 7, loglik= -1406307.819693, dloglik=0.000548 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 8, loglik= -1405752.637775, dloglik=0.000395 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 9, loglik= -1405322.992407, dloglik=0.000306 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 10, loglik= -1404988.226945, dloglik=0.000238 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 11, loglik= -1404720.087996, dloglik=0.000191 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 12, loglik= -1404502.802752, dloglik=0.000155 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 13, loglik= -1404323.839289, dloglik=0.000127 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 14, loglik= -1404174.552903, dloglik=0.000106 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 15, loglik= -1404048.247071, dloglik=0.000090 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 16, loglik= -1403940.001057, dloglik=0.000077 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 17, loglik= -1403846.063585, dloglik=0.000067 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 18, loglik= -1403763.627462, dloglik=0.000059 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 19, loglik= -1403690.551094, dloglik=0.000052 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 20, loglik= -1403625.203404, dloglik=0.000047 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 21, loglik= -1403566.323442, dloglik=0.000042 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 22, loglik= -1403512.931337, dloglik=0.000038 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 23, loglik= -1403464.255770, dloglik=0.000035 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 24, loglik= -1403419.684579, dloglik=0.000032 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 25, loglik= -1403378.725655, dloglik=0.000029</span>
<span class="va">CosMx_subset</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 960 features across 2814 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (960 features, 960 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: fast</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<p>In the following, we show how to find the signature genes based on comebeddings. First, we calculate the distance matrix.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pdistance.html">pdistance</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"fast"</span><span class="op">)</span>
<span class="co">#&gt; Calculate co-embedding distance...</span></code></pre></div>
<p>Next, we find the signature genes for each cell type</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;       B-cell  endothelial   epithelial   fibroblast   macrophage         mast </span>
<span class="co">#&gt;          193          587          226          224          220           97 </span>
<span class="co">#&gt;          mDC     monocyte   neutrophil           NK          pDC  plasmablast </span>
<span class="co">#&gt;           68           16          146           81          131          249 </span>
<span class="co">#&gt; T CD4 memory  T CD4 naive T CD8 memory  T CD8 naive         Treg     tumor 12 </span>
<span class="co">#&gt;           17          301            3           65           13           16 </span>
<span class="co">#&gt;      tumor 5      tumor 9 </span>
<span class="co">#&gt;          160            1</span>
<span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">CosMx_subset</span><span class="op">$</span><span class="va">cell_type</span>
<span class="va">df_sig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find.signature.genes.html">find.signature.genes</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">df_sig_list</span><span class="op">)</span>
<span class="co">#&gt; List of 20</span>
<span class="co">#&gt;  $ B-cell      :'data.frame':    317 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:317] 15.1 15.6 16 16 16.1 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:317] 0.539 0.311 0.347 0.601 0.399 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:317] "B-cell" "B-cell" "B-cell" "B-cell" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:317] "MS4A1" "TNFRSF13B" "CD19" "IGHM" ...</span>
<span class="co">#&gt;  $ endothelial :'data.frame':    329 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:329] 16.3 16.4 16.5 16.5 16.6 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:329] 0.661 0.44 0.283 0.307 0.641 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:329] "endothelial" "endothelial" "endothelial" "endothelial" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:329] "VWF" "CDH5" "ADGRL2" "CLEC14A" ...</span>
<span class="co">#&gt;  $ epithelial  :'data.frame':    306 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:306] 18.5 18.5 18.6 19 19.1 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:306] 0.708 0.531 0.478 0.403 0.292 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:306] "epithelial" "epithelial" "epithelial" "epithelial" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:306] "LTF" "DMBT1" "LAMP3" "ITGB6" ...</span>
<span class="co">#&gt;  $ fibroblast  :'data.frame':    327 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:327] 17.7 17.8 17.8 17.8 17.9 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:327] 0.799 0.33 0.393 0.353 0.679 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:327] "fibroblast" "fibroblast" "fibroblast" "fibroblast" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:327] "COL1A2" "COL5A2" "COL14A1" "CDH11" ...</span>
<span class="co">#&gt;  $ macrophage  :'data.frame':    306 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:306] 16.6 16.9 17 17.2 17.4 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:306] 0.709 0.659 0.832 0.523 0.305 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:306] "macrophage" "macrophage" "macrophage" "macrophage" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:306] "GPNMB" "CD163" "C1QB" "CCL18" ...</span>
<span class="co">#&gt;  $ mast        :'data.frame':    321 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:321] 16.7 16.9 17.3 17.7 18.6 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:321] 0.887 0.577 0.99 0.979 0.701 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:321] "mast" "mast" "mast" "mast" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:321] "CPA3" "IL1RL1" "TPSB2" "TPSAB1" ...</span>
<span class="co">#&gt;  $ mDC         :'data.frame':    330 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:330] 16.3 16.5 16.6 16.6 16.6 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:330] 0.721 0.779 0.382 0.382 0.235 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:330] "mDC" "mDC" "mDC" "mDC" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:330] "HLA.DQA1" "HLA.DPB1" "C1QB" "CD14" ...</span>
<span class="co">#&gt;  $ monocyte    :'data.frame':    361 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:361] 16 16.3 16.3 16.5 16.6 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:361] 0.375 0.312 0.188 0.188 0.375 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:361] "monocyte" "monocyte" "monocyte" "monocyte" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:361] "S100A8" "CSF3R" "CLEC7A" "TLR2" ...</span>
<span class="co">#&gt;  $ neutrophil  :'data.frame':    124 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:124] 17 19.4 19.7 19.9 20.1 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:124] 0.575 0.192 0.171 0.11 0.116 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:124] "neutrophil" "neutrophil" "neutrophil" "neutrophil" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:124] "CXCL8" "IL1R2" "HCAR2" "CXCR1" ...</span>
<span class="co">#&gt;  $ NK          :'data.frame':    253 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:253] 16.5 16.6 16.9 17 17.2 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:253] 0.765 0.605 0.309 0.37 0.247 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:253] "NK" "NK" "NK" "NK" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:253] "CCL5" "KLRK1" "GZMA" "NKG7" ...</span>
<span class="co">#&gt;  $ pDC         :'data.frame':    333 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:333] 16.9 17 17 17.1 17.2 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:333] 0.656 0.687 0.634 0.282 0.733 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:333] "pDC" "pDC" "pDC" "pDC" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:333] "HLA.DQA1" "C1QC" "C1QB" "CSF1R" ...</span>
<span class="co">#&gt;  $ plasmablast :'data.frame':    290 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:290] 14.7 15.2 15.8 16 16.2 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:290] 0.699 0.703 0.357 0.683 0.884 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:290] "plasmablast" "plasmablast" "plasmablast" "plasmablast" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:290] "MZB1" "JCHAIN" "CD79A" "IGHG2" ...</span>
<span class="co">#&gt;  $ T CD4 memory:'data.frame':    352 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:352] 16.6 16.6 16.7 16.8 16.8 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:352] 0.235 0.706 0.588 0.118 0.471 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:352] "T CD4 memory" "T CD4 memory" "T CD4 memory" "T CD4 memory" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:352] "FOXP3" "DCN" "COL3A1" "GZMA" ...</span>
<span class="co">#&gt;  $ T CD4 naive :'data.frame':    282 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:282] 14.7 14.9 15 15.2 15.3 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:282] 0.615 0.346 0.316 0.355 0.249 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:282] "T CD4 naive" "T CD4 naive" "T CD4 naive" "T CD4 naive" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:282] "IL7R" "CD2" "ITK" "FYB1" ...</span>
<span class="co">#&gt;  $ T CD8 memory:'data.frame':    267 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:267] 16 16.1 16.4 16.5 16.5 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:267] 0.333 0.333 0.333 0.333 0.333 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:267] "T CD8 memory" "T CD8 memory" "T CD8 memory" "T CD8 memory" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:267] "XCL2" "IL2RB" "CD84" "CCL5" ...</span>
<span class="co">#&gt;  $ T CD8 naive :'data.frame':    263 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:263] 14.8 14.9 15.1 15.3 15.3 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:263] 0.662 0.462 0.2 0.354 0.292 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:263] "T CD8 naive" "T CD8 naive" "T CD8 naive" "T CD8 naive" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:263] "CCL5" "KLRK1" "GZMA" "CD2" ...</span>
<span class="co">#&gt;  $ Treg        :'data.frame':    151 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:151] 16.3 16.3 16.4 16.5 16.5 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:151] 0.231 0.154 0.154 0.154 0.308 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:151] "Treg" "Treg" "Treg" "Treg" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:151] "IL7R" "SPOCK2" "SELL" "CD2" ...</span>
<span class="co">#&gt;  $ tumor 12    :'data.frame':    187 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:187] 19.6 19.9 20 20.1 20.2 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:187] 0.375 0.938 0.125 0.125 0.125 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:187] "tumor 12" "tumor 12" "tumor 12" "tumor 12" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:187] "UBE2C" "MZT2A" "PTHLH" "NRIP3" ...</span>
<span class="co">#&gt;  $ tumor 5     :'data.frame':    399 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:399] 15.9 16.7 17.8 18.4 18.4 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:399] 0.75 0.906 0.775 0.719 0.744 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:399] "tumor 5" "tumor 5" "tumor 5" "tumor 5" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:399] "PSCA" "CEACAM6" "CXCL5" "TACSTD2" ...</span>
<span class="co">#&gt;  $ tumor 9     :'data.frame':    24 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:24] 19.2 19.3 19.5 19.5 19.8 ...</span>
<span class="co">#&gt;   ..$ expr.prop: logi [1:24] TRUE TRUE TRUE TRUE TRUE TRUE ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:24] "tumor 9" "tumor 9" "tumor 9" "tumor 9" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:24] "MALAT1" "SLC2A4" "HDAC4" "HLA.C" ...</span></code></pre></div>
<p>Then, we obtain the top five signature genes and organize them into a data.frame. Next, we calculate the UMAP projections of coembeddings. The colname <code>distance</code> means the distance between gene (i.e., MS4A1) and cells with the specific cell type (i.e., B cell), which is calculated based on the coembedding of genes and cells in the coembedding space. The distance is smaller, the association between gene and the cell type is stronger. The colname <code>expr.prop</code> represents the expression proportion of the gene (i.e., MS4A1) within the cell type (i.e., B cell). The colname <code>label</code> means the cell types and colname <code>gene</code> denotes the gene name. By the data.frame object, we know <code>MS4A1</code> is the one of the top signature gene of B cell.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get.top.signature.dat.html">get.top.signature.dat</a></span><span class="op">(</span><span class="va">df_sig_list</span>, ntop <span class="op">=</span> <span class="fl">2</span>, expr.prop.cutoff <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>
<span class="co">#&gt;           distance expr.prop       label      gene</span>
<span class="co">#&gt; MS4A1     15.05420 0.5388601      B-cell     MS4A1</span>
<span class="co">#&gt; TNFRSF13B 15.57798 0.3108808      B-cell TNFRSF13B</span>
<span class="co">#&gt; VWF       16.28960 0.6609881 endothelial       VWF</span>
<span class="co">#&gt; CDH5      16.36390 0.4395230 endothelial      CDH5</span>
<span class="co">#&gt; LTF       18.45671 0.7079646  epithelial       LTF</span>
<span class="co">#&gt; DMBT1     18.48114 0.5309735  epithelial     DMBT1</span></code></pre></div>
<p>Next, we calculate the UMAP projections of coembeddings of cells and the selected signature genes.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembedding_umap.html">coembedding_umap</a></span><span class="op">(</span>
  <span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"fast"</span>, reduction.name <span class="op">=</span> <span class="st">"UMAP"</span>,
  gene.set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Furthermore, we visualize the cells and top two signature genes of tumor 5 in the UMAP space of coembedding. We observe that the UMAP projections of the two signature genes are near to B cells, which indicates these genes are enriched in B cells.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## choose beutifual colors</span>
<span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="fu">PRECAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Blink 23"</span>, n_colors <span class="op">=</span> <span class="fl">21</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembed_plot.html">coembed_plot</a></span><span class="op">(</span>
   <span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span>,
   gene_txtdata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">label</span><span class="op">==</span><span class="st">'tumor 5'</span><span class="op">)</span>,
   cols<span class="op">=</span><span class="va">cols_cluster</span>, pt_text_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-12-2.png" width="768"></p>
<p>Then, we visualize the cells and top two signature genes of all involved cell types in the UMAP space of coembedding. We observe that the UMAP projections of the signature genes are near to the corresponding cell type, which indicates these genes are enriched in the corresponding cells.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembed_plot.html">coembed_plot</a></span><span class="op">(</span>
   <span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span>,
   gene_txtdata <span class="op">=</span> <span class="va">dat</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, 
   pt_text_size <span class="op">=</span> <span class="fl">3</span>, alpha<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>
<span class="va">p2</span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-13-1.png" width="864"></p>
<p>In addtion, we can fully take advantages of the visualization functions in <code>Seurat</code> package for visualization. The following is an example that visualizes the cell types on the UMAP space.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_type</span> <span class="op">&lt;-</span> <span class="va">cols_cluster</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols_type</span><span class="op">)</span><span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">'UMAP'</span>, cols<span class="op">=</span><span class="va">cols_type</span><span class="op">)</span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-14-1.png" width="864"></p>
<p>Then, there is another example that we plot the first two signature genes of Tumor 5 on UMAP space, in which we observed the high expression in B cells in constrast to other cell types.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">'UMAP'</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PSCA"</span>, <span class="st">"CEACAM6"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 22621)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                              </span>
<span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] purrr_0.3.4        sp_1.5-0           SeuratObject_4.1.0 Seurat_4.1.1      </span>
<span class="co">#&gt; [5] ProFAST_1.4        gtools_3.9.2.2    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] rappdirs_0.3.3              scattermore_0.8            </span>
<span class="co">#&gt;   [3] ggthemes_4.2.4              GiRaF_1.0.1                </span>
<span class="co">#&gt;   [5] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">#&gt;   [7] ggplot2_3.4.1               bit64_4.0.5                </span>
<span class="co">#&gt;   [9] knitr_1.37                  irlba_2.3.5                </span>
<span class="co">#&gt;  [11] DelayedArray_0.20.0         data.table_1.14.2          </span>
<span class="co">#&gt;  [13] rpart_4.1.16                KEGGREST_1.34.0            </span>
<span class="co">#&gt;  [15] RCurl_1.98-1.6              generics_0.1.2             </span>
<span class="co">#&gt;  [17] BiocGenerics_0.40.0         org.Mm.eg.db_3.14.0        </span>
<span class="co">#&gt;  [19] ScaledMatrix_1.2.0          cowplot_1.1.1              </span>
<span class="co">#&gt;  [21] RSQLite_2.2.10              RANN_2.6.1                 </span>
<span class="co">#&gt;  [23] future_1.26.1               bit_4.0.4                  </span>
<span class="co">#&gt;  [25] spatstat.data_3.0-0         xml2_1.3.3                 </span>
<span class="co">#&gt;  [27] httpuv_1.6.5                SummarizedExperiment_1.24.0</span>
<span class="co">#&gt;  [29] assertthat_0.2.1            viridis_0.6.2              </span>
<span class="co">#&gt;  [31] xfun_0.29                   hms_1.1.1                  </span>
<span class="co">#&gt;  [33] jquerylib_0.1.4             evaluate_0.15              </span>
<span class="co">#&gt;  [35] promises_1.2.0.1            fansi_1.0.4                </span>
<span class="co">#&gt;  [37] progress_1.2.2              dbplyr_2.1.1               </span>
<span class="co">#&gt;  [39] igraph_1.3.5                DBI_1.1.2                  </span>
<span class="co">#&gt;  [41] htmlwidgets_1.5.4           spatstat.geom_2.4-0        </span>
<span class="co">#&gt;  [43] stats4_4.1.2                ellipsis_0.3.2             </span>
<span class="co">#&gt;  [45] RSpectra_0.16-1             dplyr_1.0.9                </span>
<span class="co">#&gt;  [47] ggpubr_0.4.0                backports_1.4.1            </span>
<span class="co">#&gt;  [49] DR.SC_3.4                   insight_0.17.0             </span>
<span class="co">#&gt;  [51] biomaRt_2.50.3              deldir_1.0-6               </span>
<span class="co">#&gt;  [53] sparseMatrixStats_1.6.0     MatrixGenerics_1.6.0       </span>
<span class="co">#&gt;  [55] vctrs_0.6.1                 SingleCellExperiment_1.16.0</span>
<span class="co">#&gt;  [57] Biobase_2.54.0              ROCR_1.0-11                </span>
<span class="co">#&gt;  [59] abind_1.4-5                 cachem_1.0.6               </span>
<span class="co">#&gt;  [61] withr_2.5.0                 PRECAST_1.6.3              </span>
<span class="co">#&gt;  [63] progressr_0.10.1            sctransform_0.3.3          </span>
<span class="co">#&gt;  [65] prettyunits_1.1.1           mclust_5.4.10              </span>
<span class="co">#&gt;  [67] goftest_1.2-3               cluster_2.1.2              </span>
<span class="co">#&gt;  [69] lazyeval_0.2.2              crayon_1.5.1               </span>
<span class="co">#&gt;  [71] pkgconfig_2.0.3             labeling_0.4.2             </span>
<span class="co">#&gt;  [73] GenomeInfoDb_1.30.1         nlme_3.1-155               </span>
<span class="co">#&gt;  [75] vipor_0.4.5                 nnet_7.3-17                </span>
<span class="co">#&gt;  [77] rlang_1.1.0                 globals_0.15.0             </span>
<span class="co">#&gt;  [79] lifecycle_1.0.3             miniUI_0.1.1.1             </span>
<span class="co">#&gt;  [81] filelock_1.0.2              BiocFileCache_2.2.1        </span>
<span class="co">#&gt;  [83] rsvd_1.0.5                  rprojroot_2.0.3            </span>
<span class="co">#&gt;  [85] polyclip_1.10-0             matrixStats_0.62.0         </span>
<span class="co">#&gt;  [87] lmtest_0.9-40               Matrix_1.4-0               </span>
<span class="co">#&gt;  [89] carData_3.0-5               zoo_1.8-10                 </span>
<span class="co">#&gt;  [91] beeswarm_0.4.0              ggridges_0.5.3             </span>
<span class="co">#&gt;  [93] png_0.1-7                   viridisLite_0.4.1          </span>
<span class="co">#&gt;  [95] bitops_1.0-7                KernSmooth_2.23-20         </span>
<span class="co">#&gt;  [97] Biostrings_2.62.0           blob_1.2.2                 </span>
<span class="co">#&gt;  [99] DelayedMatrixStats_1.16.0   stringr_1.4.0              </span>
<span class="co">#&gt; [101] parallelly_1.32.0           spatstat.random_2.2-0      </span>
<span class="co">#&gt; [103] rstatix_0.7.0               S4Vectors_0.32.3           </span>
<span class="co">#&gt; [105] ggsignif_0.6.3              beachmat_2.10.0            </span>
<span class="co">#&gt; [107] scales_1.2.1                memoise_2.0.1              </span>
<span class="co">#&gt; [109] magrittr_2.0.3              plyr_1.8.7                 </span>
<span class="co">#&gt; [111] ica_1.0-2                   zlibbioc_1.40.0            </span>
<span class="co">#&gt; [113] compiler_4.1.2              RColorBrewer_1.1-3         </span>
<span class="co">#&gt; [115] fitdistrplus_1.1-8          cli_3.2.0                  </span>
<span class="co">#&gt; [117] XVector_0.34.0              listenv_0.8.0              </span>
<span class="co">#&gt; [119] patchwork_1.1.1             pbapply_1.5-0              </span>
<span class="co">#&gt; [121] MASS_7.3-55                 mgcv_1.8-39                </span>
<span class="co">#&gt; [123] tidyselect_1.1.2            stringi_1.7.6              </span>
<span class="co">#&gt; [125] textshaping_0.3.6           highr_0.9                  </span>
<span class="co">#&gt; [127] yaml_2.3.6                  BiocSingular_1.10.0        </span>
<span class="co">#&gt; [129] ggrepel_0.9.1               grid_4.1.2                 </span>
<span class="co">#&gt; [131] sass_0.4.1                  tools_4.1.2                </span>
<span class="co">#&gt; [133] future.apply_1.9.0          parallel_4.1.2             </span>
<span class="co">#&gt; [135] rstudioapi_0.13             gridExtra_2.3              </span>
<span class="co">#&gt; [137] farver_2.1.1                Rtsne_0.16                 </span>
<span class="co">#&gt; [139] digest_0.6.29               rgeos_0.5-9                </span>
<span class="co">#&gt; [141] FNN_1.1.3.1                 shiny_1.7.1                </span>
<span class="co">#&gt; [143] Rcpp_1.0.10                 GenomicRanges_1.46.1       </span>
<span class="co">#&gt; [145] car_3.0-12                  broom_0.7.12               </span>
<span class="co">#&gt; [147] scuttle_1.4.0               performance_0.8.0          </span>
<span class="co">#&gt; [149] later_1.3.0                 harmony_0.1.0              </span>
<span class="co">#&gt; [151] RcppAnnoy_0.0.19            org.Hs.eg.db_3.14.0        </span>
<span class="co">#&gt; [153] httr_1.4.3                  AnnotationDbi_1.56.2       </span>
<span class="co">#&gt; [155] colorspace_2.1-0            XML_3.99-0.9               </span>
<span class="co">#&gt; [157] fs_1.5.2                    tensor_1.5                 </span>
<span class="co">#&gt; [159] reticulate_1.25             IRanges_2.28.0             </span>
<span class="co">#&gt; [161] splines_4.1.2               uwot_0.1.11                </span>
<span class="co">#&gt; [163] spatstat.utils_3.0-1        pkgdown_2.0.6              </span>
<span class="co">#&gt; [165] scater_1.25.1               plotly_4.10.0              </span>
<span class="co">#&gt; [167] systemfonts_1.0.4           xtable_1.8-4               </span>
<span class="co">#&gt; [169] jsonlite_1.8.0              R6_2.5.1                   </span>
<span class="co">#&gt; [171] pillar_1.9.0                htmltools_0.5.2            </span>
<span class="co">#&gt; [173] mime_0.12                   tictoc_1.2                 </span>
<span class="co">#&gt; [175] glue_1.6.2                  fastmap_1.1.0              </span>
<span class="co">#&gt; [177] BiocParallel_1.28.3         BiocNeighbors_1.12.0       </span>
<span class="co">#&gt; [179] codetools_0.2-18            furrr_0.3.1                </span>
<span class="co">#&gt; [181] utf8_1.2.3                  lattice_0.20-45            </span>
<span class="co">#&gt; [183] bslib_0.3.1                 spatstat.sparse_2.1-1      </span>
<span class="co">#&gt; [185] tibble_3.2.1                curl_4.3.2                 </span>
<span class="co">#&gt; [187] ggbeeswarm_0.6.0            leiden_0.4.2               </span>
<span class="co">#&gt; [189] survival_3.2-13             CompQuadForm_1.4.3         </span>
<span class="co">#&gt; [191] rmarkdown_2.11              desc_1.4.0                 </span>
<span class="co">#&gt; [193] munsell_0.5.0               GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt; [195] reshape2_1.4.4              gtable_0.3.3               </span>
<span class="co">#&gt; [197] spatstat.core_2.4-4</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
