<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>FAST: single DLPFC section • ProFAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="FAST: single DLPFC section">
<meta property="og:description" content="ProFAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ProFAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CosMx.html">CoFAST: NSCLC CosMx data coembedding</a>
    </li>
    <li>
      <a href="../articles/FASTdlpfc.html">FAST: single DLPFC section</a>
    </li>
    <li>
      <a href="../articles/FASTdlpfc2.html">FAST: two DLPFC sections</a>
    </li>
    <li>
      <a href="../articles/FASTsimu.html">FAST: simulation</a>
    </li>
    <li>
      <a href="../articles/pbmc3k.html">CoFAST: PBMC scRNA-seq data coembedding</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/ProFAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>FAST: single DLPFC section</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2024-03-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/ProFAST/blob/HEAD/vignettes/FASTdlpfc.Rmd" class="external-link"><code>vignettes/FASTdlpfc.Rmd</code></a></small>
      <div class="hidden name"><code>FASTdlpfc.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the FAST workflow for the analysis of single-section rather than multi-secitons data, one humn dorsolateral prefrontal cortex (DLPFC) spatial transcriptomics dataset. In this vignette, the workflow of FAST consists of three steps</p>
<ul>
<li>Independent preprocessing</li>
<li>Spatial dimension reduction using FAST</li>
<li>Downstream analysis (i.e. , embedding alignment and spatial clustering using iSC-MEB, visualization of clusters and embeddings, remove unwanted variation, differential expression analysis)</li>
</ul>
<p>We demonstrate the use of FAST to one DLPFC Visium data that are <a href="https://github.com/feiyoung/ProFAST/tree/main/vignettes_data" class="external-link">here</a>, which can be downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/FAST/blob/main/vignettes_data/seulist2_ID9_10.RDS?raw=true"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">githubURL</span>,<span class="st">"seulist2_ID9_10.RDS"</span>,mode<span class="op">=</span><span class="st">'wb'</span><span class="op">)</span></code></pre></div>
<p>Then load to R. Here, we only focus one section.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dlpfc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./seulist2_ID9_10.RDS"</span><span class="op">)</span>
<span class="va">dlpfc</span> <span class="op">&lt;-</span> <span class="va">dlpfc2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/ProFAST" class="external-link">ProFAST</a></span><span class="op">)</span> <span class="co"># load the package of FAST method</span>
<span class="co">#&gt; Loading required package: gtools</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; PRECAST :  An efficient data integration method is provided for multiple spatial transcriptomics data with non-cluster-relevant effects such as the complex batch effects. It unifies spatial factor analysis simultaneously with spatial clustering and embedding alignment, requiring only partially shared cell/domain clusters across datasets. More details can be referred to Wei Liu, et al. (2023) &lt;doi:10.1038/s41467-023-35947-w&gt;.   Check out our Package website (https://feiyoung.github.io/PRECAST/index.html) for a more complete description of the methods and analyses</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'Seurat' was built under R version 4.1.3</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; Attaching sp</span></code></pre></div>
<div class="section level2">
<h2 id="view-the-dlpfc-data">View the DLPFC data<a class="anchor" aria-label="anchor" href="#view-the-dlpfc-data"></a>
</h2>
<p>First, we view the the spatial transcriptomics data with Visium platform. There are ~15000 genes and ~3600 spots.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dlpfc</span> <span class="co">## a list including three Seurat object with default assay: RNA</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 15027 features across 3639 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (15027 features, 2000 variable features)</span></code></pre></div>
<p>We observed that the genes are Ensembl IDs. In the following, we will transfer the Ensembl IDs to gene symbols for matching the housekeeping genes in the downstream analysis for removing the unwanted variations.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">dlpfc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;  [1] "ENSG00000237491" "ENSG00000230368" "ENSG00000187634" "ENSG00000188976"</span>
<span class="co">#&gt;  [5] "ENSG00000187961" "ENSG00000272512" "ENSG00000188290" "ENSG00000187608"</span>
<span class="co">#&gt;  [9] "ENSG00000188157" "ENSG00000131591"</span>
<span class="va">count</span> <span class="op">&lt;-</span> <span class="va">dlpfc</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="fu"><a href="../reference/transferGeneNames.html">transferGeneNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>, now_name <span class="op">=</span> <span class="st">"ensembl"</span>,
                                                 to_name<span class="op">=</span><span class="st">"symbol"</span>,
                                                 species<span class="op">=</span><span class="st">"Human"</span>, Method<span class="op">=</span><span class="st">'eg.db'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;  [1] "LINC01409"       "FAM41C"          "SAMD11"          "NOC2L"          </span>
<span class="co">#&gt;  [5] "KLHL17"          "ENSG00000272512" "HES4"            "ISG15"          </span>
<span class="co">#&gt;  [9] "AGRN"            "C1ORF159"</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count</span>, meta.data <span class="op">=</span> <span class="va">dlpfc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span>
<span class="co">#&gt; Warning: Non-unique features (rownames) present in the input matrix, making</span>
<span class="co">#&gt; unique</span>
<span class="va">seu</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 15027 features across 3639 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (15027 features, 0 variable features)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>We show how to preprocessing before fitting FAST, including log-normalization (if user use the gaussian version of FAST), and select highly variable genes.</p>
<ul>
<li>Note: the spatial coordinates must be contained in the meta data and named as <code>row</code> and <code>col</code>, which benefits the identification of spaital coordinates by FAST.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Check the spatial coordinates: they are named as "row" and "col"!</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seu</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                       orig.ident nCount_RNA nFeature_RNA            barcode</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1 SeuratProject       8436         3564 AAACAAGTATCTCCCA-1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1 SeuratProject       1662         1145 AAACAATCTACTAGCA-1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1 SeuratProject       3758         1949 AAACACCAATAACTGC-1</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1 SeuratProject       5418         2409 AAACAGAGCGACTCCT-1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1 SeuratProject       4262         2248 AAACAGCTTTCAGAAG-1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1 SeuratProject       3984         2159 AAACAGGGTCTATATT-1</span>
<span class="co">#&gt;                    sample_name tissue row col imagerow imagecol Cluster height</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1      151673      1  50 102 381.0981 440.6391       7    600</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1      151673      1   3  43 126.3276 259.6310       4    600</span>
<span class="co">#&gt; AAACACCAATAACTGC-1      151673      1  59  19 427.7678 183.0783       8    600</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1      151673      1  14  94 186.8137 417.2367       6    600</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1      151673      1  43   9 341.2691 152.7003       3    600</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1      151673      1  47  13 362.9163 164.9415       3    600</span>
<span class="co">#&gt;                    width sum_umi sum_gene subject position replicate</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1   600    8458     3586  Br8100        0         1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1   600    1667     1150  Br8100        0         1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1   600    3769     1960  Br8100        0         1</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1   600    5433     2424  Br8100        0         1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1   600    4278     2264  Br8100        0         1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1   600    4004     2178  Br8100        0         1</span>
<span class="co">#&gt;                    subject_position discard                       key</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1      Br8100_pos0   FALSE 151673_AAACAAGTATCTCCCA-1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1      Br8100_pos0   FALSE 151673_AAACAATCTACTAGCA-1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1      Br8100_pos0   FALSE 151673_AAACACCAATAACTGC-1</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1      Br8100_pos0   FALSE 151673_AAACAGAGCGACTCCT-1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1      Br8100_pos0   FALSE 151673_AAACAGCTTTCAGAAG-1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1      Br8100_pos0   FALSE 151673_AAACAGGGTCTATATT-1</span>
<span class="co">#&gt;                    cell_count SNN_k50_k4 SNN_k50_k5 SNN_k50_k6 SNN_k50_k7</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1          6          1          1          1          1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1         16          1          1          1          1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1          5          2          3          4          5</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1          2          1          1          1          1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1          4          1          1          1          1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1          6          2          4          5          6</span>
<span class="co">#&gt;                    SNN_k50_k8 SNN_k50_k9 SNN_k50_k10 SNN_k50_k11 SNN_k50_k12</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1          1          1           1           4           4</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1          1          1           1           4           4</span>
<span class="co">#&gt; AAACACCAATAACTGC-1          8          9          10          11          10</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1          1          1           1           2           2</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1          1          3           3           3           3</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1          5          6           7           8          12</span>
<span class="co">#&gt;                    SNN_k50_k13 SNN_k50_k14 SNN_k50_k15 SNN_k50_k16 SNN_k50_k17</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1           4           4           3           3           3</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1           4           4           3           3           3</span>
<span class="co">#&gt; AAACACCAATAACTGC-1           9          10          11          11          12</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1           2           2           1           1           1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1           3           3           2           2           2</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1          12          13          14          15          16</span>
<span class="co">#&gt;                    SNN_k50_k18 SNN_k50_k19 SNN_k50_k20 SNN_k50_k21 SNN_k50_k22</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1           2           2           2           2           1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1           2           2           2           2           1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1          13          12          11          11          12</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1           4           4           4           4           3</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1           1           1           1           1           6</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1          17          17          17          16          17</span>
<span class="co">#&gt;                    SNN_k50_k23 SNN_k50_k24 SNN_k50_k25 SNN_k50_k26 SNN_k50_k27</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1           1           1           6           5           5</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1           1           1          10          10          10</span>
<span class="co">#&gt; AAACACCAATAACTGC-1          20          20          21          22          21</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1           3           3           2           2           2</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1           6           6           5           4           4</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1          16          16          17          18          17</span>
<span class="co">#&gt;                    SNN_k50_k28 GraphBased Maynard Martinowich Layer layer_guess</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1           4          7     2_3           2    NA      Layer3</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1          10          4     2_3           3    NA      Layer1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1          22          8      WM          WM    NA          WM</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1           7          6       6           6    NA      Layer3</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1           3          3       5           5    NA      Layer5</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1          18          3       5           5    NA      Layer6</span>
<span class="co">#&gt;                    layer_guess_reordered layer_guess_reordered_short expr_chrM</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1                Layer3                          L3      1407</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1                Layer1                          L1       204</span>
<span class="co">#&gt; AAACACCAATAACTGC-1                    WM                          WM       430</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1                Layer3                          L3      1316</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1                Layer5                          L5       651</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1                Layer6                          L6       621</span>
<span class="co">#&gt;                    expr_chrM_ratio SpatialDE_PCA SpatialDE_pool_PCA HVG_PCA</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1       0.1663514             3                  3       2</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1       0.1223755             2                  5       3</span>
<span class="co">#&gt; AAACACCAATAACTGC-1       0.1140886             4                  4       7</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1       0.2422234             3                  3       2</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1       0.1521739             2                  1       1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1       0.1550949             6                  7       8</span>
<span class="co">#&gt;                    pseudobulk_PCA markers_PCA SpatialDE_UMAP</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1              2           2              1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1              7           8              2</span>
<span class="co">#&gt; AAACACCAATAACTGC-1              5           5              5</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1              3           6              1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1              4           2              1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1              8           4              1</span>
<span class="co">#&gt;                    SpatialDE_pool_UMAP HVG_UMAP pseudobulk_UMAP markers_UMAP</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1                   1        3               1            1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1                   2        2               2            1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1                   5        6               6            1</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1                   1        3               1            3</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1                   1        7               3            4</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1                   7        1               3            4</span>
<span class="co">#&gt;                    SpatialDE_PCA_spatial SpatialDE_pool_PCA_spatial</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1                     3                          1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1                     7                          5</span>
<span class="co">#&gt; AAACACCAATAACTGC-1                     5                          4</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1                     3                          3</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1                     2                          1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1                     6                          8</span>
<span class="co">#&gt;                    HVG_PCA_spatial pseudobulk_PCA_spatial markers_PCA_spatial</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1               1                      3                   1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1               2                      2                   3</span>
<span class="co">#&gt; AAACACCAATAACTGC-1               4                      5                   3</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1               1                      2                   2</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1               2                      4                   1</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1               7                      7                   7</span>
<span class="co">#&gt;                    SpatialDE_UMAP_spatial SpatialDE_pool_UMAP_spatial</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1                      7                           1</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1                      2                           1</span>
<span class="co">#&gt; AAACACCAATAACTGC-1                      5                           7</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1                      3                           4</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1                      3                           3</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1                      3                           3</span>
<span class="co">#&gt;                    HVG_UMAP_spatial pseudobulk_UMAP_spatial</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1                1                       2</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1                4                       2</span>
<span class="co">#&gt; AAACACCAATAACTGC-1                5                       3</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1                2                       1</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1                8                       4</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1                3                       4</span>
<span class="co">#&gt;                    markers_UMAP_spatial sizeFactor x_gmm     r_gmm.V1</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1                    1  2.6257715     1 6.148291e-01</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1                    3  0.5269265     4 9.639893e-04</span>
<span class="co">#&gt; AAACACCAATAACTGC-1                    2  1.0484921     5 2.420481e-59</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1                    1  1.4784490     1 9.126456e-01</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1                    4  1.3386770     3 4.862401e-04</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1                    4  1.1790057     7 5.331015e-12</span>
<span class="co">#&gt;                        r_gmm.V2     r_gmm.V3     r_gmm.V4     r_gmm.V5</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1 2.720022e-24 3.373629e-03 4.191687e-05 1.159537e-49</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1 9.387657e-21 7.717454e-04 9.916840e-01 2.504924e-42</span>
<span class="co">#&gt; AAACACCAATAACTGC-1 3.239627e-12 2.125771e-51 4.454654e-56 1.000000e+00</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1 1.816712e-24 5.985415e-04 3.034282e-05 1.093182e-49</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1 1.298154e-13 8.302244e-01 1.998361e-10 2.101466e-36</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1 1.795491e-03 4.993974e-05 7.770341e-16 1.156953e-19</span>
<span class="co">#&gt;                        r_gmm.V6     r_gmm.V7 x_icm    r_icm.V1    r_icm.V2</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1 3.817553e-01 1.493637e-08     6 0.216099294 0.029245859</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1 6.580215e-03 5.273170e-08     4 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACACCAATAACTGC-1 7.001587e-58 7.765638e-38     5 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1 8.672553e-02 1.234249e-08     6 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1 5.664366e-03 1.636250e-01     2 0.007702348 0.935915561</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1 1.319611e-11 9.981546e-01     2 0.002442427 0.985345437</span>
<span class="co">#&gt;                       r_icm.V3    r_icm.V4    r_icm.V5    r_icm.V6    r_icm.V7</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1 0.029245859 0.079498487 0.029245859 0.587418783 0.029245859</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1 0.002442427 0.985345437 0.002442427 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACACCAATAACTGC-1 0.002442427 0.002442427 0.985345437 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1 0.002442427 0.002442427 0.002442427 0.985345437 0.002442427</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1 0.025572697 0.007702348 0.007702348 0.007702348 0.007702348</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1 0.002442427 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">#&gt;                    x_gc     r_gc.V1     r_gc.V2     r_gc.V3     r_gc.V4</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1    6 0.216099294 0.029245859 0.029245859 0.079498487</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1    4 0.002442427 0.002442427 0.002442427 0.985345437</span>
<span class="co">#&gt; AAACACCAATAACTGC-1    5 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1    6 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1    7 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1    7 0.014928224 0.110305483 0.014928224 0.014928224</span>
<span class="co">#&gt;                        r_gc.V5     r_gc.V6     r_gc.V7</span>
<span class="co">#&gt; AAACAAGTATCTCCCA-1 0.029245859 0.587418783 0.029245859</span>
<span class="co">#&gt; AAACAATCTACTAGCA-1 0.002442427 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACACCAATAACTGC-1 0.985345437 0.002442427 0.002442427</span>
<span class="co">#&gt; AAACAGAGCGACTCCT-1 0.002442427 0.985345437 0.002442427</span>
<span class="co">#&gt; AAACAGCTTTCAGAAG-1 0.002442427 0.002442427 0.985345437</span>
<span class="co">#&gt; AAACAGGGTCTATATT-1 0.014928224 0.014928224 0.815053399</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 15027 features across 3639 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (15027 features, 2000 variable features)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;  [1] "NPY"   "PLP1"  "IGKC"  "HBB"   "HBA2"  "MBP"   "SST"   "IGHG3" "CNP"  </span>
<span class="co">#&gt; [10] "CRYAB"</span></code></pre></div>
<details><p><summary><strong>Find spatially variable genes</strong></summary> Users can also use the spatially variable genes by the following command:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">DR.SC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DR.SC/man/FindSVGs.html" class="external-link">FindSVGs</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>
<span class="co">### Check the results</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="fit-fast-for-this-data">Fit FAST for this data<a class="anchor" aria-label="anchor" href="#fit-fast-for-this-data"></a>
</h2>
<p>For function <code>FAST_single</code>, users can specify the number of factors <code>q</code> and the fitted model <code>fit.model</code>. The <code>q</code> sets the number of spatial factors to be extracted, and a lareger one means more information to be extracted but higher computaional cost. The <code>fit.model</code> specifies the version of FAST to be fitted. The Gaussian version (<code>gaussian</code>) models the log-normalized matrix while the Poisson verion (<code>poisson</code>) models the count matrix; default as <code>poisson</code>. (Note: The computational time required to run the analysis on personal PCs is approximately ~0.5 minute on a personal PC.)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Adj_sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">seu</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, platform <span class="op">=</span> <span class="st">"Visium"</span><span class="op">)</span>
<span class="co">#&gt; The spatial cooridnates are 2 dimensions</span>
<span class="co">#&gt; Neighbors were identified for 3638 out of 3639 spots.</span>
<span class="co">### set q= 15 here</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FAST_single.html">FAST_single</a></span><span class="op">(</span><span class="va">seu</span>, Adj_sp<span class="op">=</span><span class="va">Adj_sp</span>, q<span class="op">=</span> <span class="fl">15</span>, fit.model<span class="op">=</span><span class="st">'poisson'</span><span class="op">)</span>
<span class="co">#&gt; ******Run the Poisson version of FAST...</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Finish variable intialization </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 2, loglik= 4721646.936916, dloglik=1.002199 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 3, loglik= 4735495.060408, dloglik=0.002933 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 4, loglik= 4738255.198282, dloglik=0.000583 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 5, loglik= 4739712.043458, dloglik=0.000307 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 6, loglik= 4740608.245939, dloglik=0.000189 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 7, loglik= 4741211.569174, dloglik=0.000127 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 8, loglik= 4741644.069050, dloglik=0.000091 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 9, loglik= 4741969.914462, dloglik=0.000069 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 10, loglik= 4742226.046132, dloglik=0.000054 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 11, loglik= 4742434.476433, dloglik=0.000044 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 12, loglik= 4742608.945866, dloglik=0.000037 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 13, loglik= 4742758.174912, dloglik=0.000031 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 14, loglik= 4742887.788831, dloglik=0.000027 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 15, loglik= 4743001.446850, dloglik=0.000024 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 16, loglik= 4743101.582934, dloglik=0.000021 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 17, loglik= 4743189.893033, dloglik=0.000019 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 18, loglik= 4743267.658362, dloglik=0.000016 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 19, loglik= 4743335.946812, dloglik=0.000014 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 20, loglik= 4743395.724873, dloglik=0.000013 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 21, loglik= 4743447.907537, dloglik=0.000011 </span>
<span class="co">#&gt; Satrt ICM and E-step! </span>
<span class="co">#&gt; Finish ICM and E-step! </span>
<span class="co">#&gt; iter = 22, loglik= 4743493.369602, dloglik=0.000010</span>
<span class="co">#&gt; 2024-03-16 19:23:14 : ***** Finish FAST, 0.481 mins elapsed.</span>
<span class="va">seu</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 15027 features across 3639 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (15027 features, 2000 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: fast</span></code></pre></div>
<details><p><summary><strong>Run possion version</strong></summary> Users can also use the gaussian version by the following command:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Adj_sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">seu</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, platform <span class="op">=</span> <span class="st">"Visium"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FAST_single.html">FAST_single</a></span><span class="op">(</span><span class="va">dlpfc</span>, Adj_sp<span class="op">=</span><span class="va">Adj_sp</span>, q<span class="op">=</span> <span class="fl">15</span>, fit.model<span class="op">=</span><span class="st">'gaussian'</span><span class="op">)</span>
<span class="co">### Check the results</span>
<span class="va">seu</span></code></pre></div>
</details><div class="section level3">
<h3 id="evaluate-the-adjusted-mcfaddens-pseudo-r-square">Evaluate the adjusted McFadden’s pseudo R-square<a class="anchor" aria-label="anchor" href="#evaluate-the-adjusted-mcfaddens-pseudo-r-square"></a>
</h3>
<p>Next, we investigate the performance of dimension reduction by calculating the adjusted McFadden’s pseudo R-square. The manual annotations are regarded as the ground truth in the <code>meta.data</code> of <code>seu</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Obtain the true labels</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">layer_guess_reordered</span>
<span class="co">### Evaluate the MacR2</span>
<span class="va">Mac</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/get_r2_mcfadden.html">get_r2_mcfadden</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seu</span>, reduction<span class="op">=</span><span class="st">'fast'</span><span class="op">)</span>, <span class="va">y</span><span class="op">)</span>
<span class="co">#&gt; # weights:  119 (96 variable)</span>
<span class="co">#&gt; initial  value 7026.681548 </span>
<span class="co">#&gt; iter  10 value 1409.672035</span>
<span class="co">#&gt; iter  20 value 1278.713363</span>
<span class="co">#&gt; iter  30 value 1180.414145</span>
<span class="co">#&gt; iter  40 value 1116.295883</span>
<span class="co">#&gt; iter  50 value 1092.963545</span>
<span class="co">#&gt; iter  60 value 1065.327354</span>
<span class="co">#&gt; iter  70 value 1015.779215</span>
<span class="co">#&gt; iter  80 value 968.708711</span>
<span class="co">#&gt; iter  90 value 935.972224</span>
<span class="co">#&gt; iter 100 value 913.623163</span>
<span class="co">#&gt; final  value 913.623163 </span>
<span class="co">#&gt; stopped after 100 iterations</span>
<span class="co">### output them</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"MacFadden's R-square of FAST is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">Mac</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "MacFadden's R-square of FAST is 0.86"</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="clustering-analysis-based-on-fast-embedding">Clustering analysis based on FAST embedding<a class="anchor" aria-label="anchor" href="#clustering-analysis-based-on-fast-embedding"></a>
</h2>
<p>Based on the embeddings from FAST, we use <code>Louvain</code> to perform clustering. In this downstream analysis, other methods for clustering can be also used.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">'fast'</span><span class="op">)</span>
<span class="co">#&gt; Computing nearest neighbor graph</span>
<span class="co">#&gt; Computing SNN</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu</span>, resolution <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 3639</span>
<span class="co">#&gt; Number of edges: 115298</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 10 random starts: 0.8645</span>
<span class="co">#&gt; Number of communities: 7</span>
<span class="co">#&gt; Elapsed time: 0 seconds</span>
<span class="va">seu</span><span class="op">$</span><span class="va">fast.cluster</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">ARI.fast</span> <span class="op">&lt;-</span> <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">seu</span><span class="op">$</span><span class="va">fast.cluster</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ARI of PCA is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">ARI.fast</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "ARI of PCA is 0.423"</span></code></pre></div>
<p>For comparison, we also run PCA to obtain PCA embeddings, and then conduct louvain clustering.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>
<span class="co">#&gt; Centering and scaling data matrix</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seu</span>, npcs<span class="op">=</span><span class="fl">15</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">Mac.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_r2_mcfadden.html">get_r2_mcfadden</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seu</span>, reduction<span class="op">=</span><span class="st">'pca'</span><span class="op">)</span>, <span class="va">y</span><span class="op">)</span>
<span class="co">#&gt; # weights:  119 (96 variable)</span>
<span class="co">#&gt; initial  value 7026.681548 </span>
<span class="co">#&gt; iter  10 value 2638.636241</span>
<span class="co">#&gt; iter  20 value 2513.930759</span>
<span class="co">#&gt; iter  30 value 2329.340597</span>
<span class="co">#&gt; iter  40 value 2069.156942</span>
<span class="co">#&gt; iter  50 value 2012.665652</span>
<span class="co">#&gt; iter  60 value 1975.876382</span>
<span class="co">#&gt; iter  70 value 1910.031745</span>
<span class="co">#&gt; iter  80 value 1844.981378</span>
<span class="co">#&gt; iter  90 value 1818.036414</span>
<span class="co">#&gt; iter 100 value 1793.559494</span>
<span class="co">#&gt; final  value 1793.559494 </span>
<span class="co">#&gt; stopped after 100 iterations</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"MacFadden's R-square of PCA is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">Mac.pca</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "MacFadden's R-square of PCA is 0.726"</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">'pca'</span>, graph.name <span class="op">=</span><span class="st">"pca.graph"</span><span class="op">)</span>
<span class="co">#&gt; Computing nearest neighbor graph</span>
<span class="co">#&gt; Computing SNN</span>
<span class="co">#&gt; Only one graph name supplied, storing nearest-neighbor graph only</span>
<span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>,graph.name <span class="op">=</span> <span class="st">'pca.graph'</span><span class="op">)</span> 
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 3639</span>
<span class="co">#&gt; Number of edges: 34702</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 10 random starts: 0.7504</span>
<span class="co">#&gt; Number of communities: 11</span>
<span class="co">#&gt; Elapsed time: 0 seconds</span>
<span class="co">#&gt; 3 singletons identified. 8 final clusters.</span>
<span class="va">seu</span><span class="op">$</span><span class="va">pca.cluster</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">ARI.pca</span> <span class="op">&lt;-</span> <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">seu</span><span class="op">$</span><span class="va">pca.cluster</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ARI of PCA is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">ARI.pca</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "ARI of PCA is 0.355"</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>First, user can choose a beautiful color schema using <code><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors()</a></code> in the R package <code>PRECAST</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Nature 10"</span>, n_colors <span class="op">=</span> <span class="fl">8</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="FASTdlpfc_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Then, we plot the spatial scatter plot for clusters using the function <code><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot()</a></code> in the R package <code>Seurat</code>. We observe that the clusters from PCA are more messy while the clusters from FAST are more smoothing in spatial coordinates.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">PRECAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/Add_embed.html" class="external-link">Add_embed</a></span><span class="op">(</span>embed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">seu</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, <span class="va">seu</span>, embed_name <span class="op">=</span> <span class="st">'Spatial'</span><span class="op">)</span>
<span class="va">seu</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 15027 features across 3639 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (15027 features, 2000 variable features)</span>
<span class="co">#&gt;  3 dimensional reductions calculated: fast, pca, Spatial</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">'Spatial'</span>, group.by <span class="op">=</span> <span class="st">'pca.cluster'</span>,cols <span class="op">=</span> <span class="va">cols_cluster</span>, pt.size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">'Spatial'</span>, group.by <span class="op">=</span> <span class="st">'fast.cluster'</span>,cols <span class="op">=</span> <span class="va">cols_cluster</span>, pt.size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/drawFigs.html" class="external-link">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>,layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span> <span class="op">)</span></code></pre></div>
<p><img src="FASTdlpfc_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<p>Next, we visualize the clusters from <code>FAST</code> on the UMAP space, and observe the clusters are well separated in general.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"fast"</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span>
<span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">#&gt; This message will be shown once per session</span>
<span class="co">#&gt; 19:23:34 UMAP embedding parameters a = 0.9922 b = 1.112</span>
<span class="co">#&gt; 19:23:34 Read 3639 rows and found 15 numeric columns</span>
<span class="co">#&gt; 19:23:34 Using Annoy for neighbor search, n_neighbors = 30</span>
<span class="co">#&gt; 19:23:34 Building Annoy index with metric = cosine, n_trees = 50</span>
<span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span>
<span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span>
<span class="co">#&gt; **************************************************|</span>
<span class="co">#&gt; 19:23:35 Writing NN index file to temp file C:\Users\LIUXIA~1\AppData\Local\Temp\RtmpCKRIph\file2fc0741f3d0a</span>
<span class="co">#&gt; 19:23:35 Searching Annoy index using 1 thread, search_k = 3000</span>
<span class="co">#&gt; 19:23:36 Annoy recall = 100%</span>
<span class="co">#&gt; 19:23:37 Commencing smooth kNN distance calibration using 1 thread</span>
<span class="co">#&gt; 19:23:39 Initializing from normalized Laplacian + noise</span>
<span class="co">#&gt; 19:23:39 Commencing optimization for 500 epochs, with 149882 positive edges</span>
<span class="co">#&gt; 19:23:51 Optimization finished</span>
<span class="va">seu</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 15027 features across 3639 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (15027 features, 2000 variable features)</span>
<span class="co">#&gt;  4 dimensional reductions calculated: fast, pca, Spatial, umap</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction<span class="op">=</span><span class="st">'umap'</span>, group.by <span class="op">=</span> <span class="st">"fast.cluster"</span><span class="op">)</span></code></pre></div>
<p><img src="FASTdlpfc_files/figure-html/unnamed-chunk-17-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="differential-epxression-analysis">Differential epxression analysis<a class="anchor" aria-label="anchor" href="#differential-epxression-analysis"></a>
</h2>
<p>Finally, we condut the differential expression (DE) analysis. The function <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers()</a></code> in the <code>Seurat</code> R package is ued to achieve this analysis. And we extract the top five DE genes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">fast.cluster</span>
<span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>
<span class="co">#&gt; Calculating cluster 0</span>
<span class="co">#&gt; Calculating cluster 1</span>
<span class="co">#&gt; Calculating cluster 2</span>
<span class="co">#&gt; Calculating cluster 3</span>
<span class="co">#&gt; Calculating cluster 4</span>
<span class="co">#&gt; Calculating cluster 5</span>
<span class="co">#&gt; Calculating cluster 6</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.1.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">5</span>
<span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top5</span>
<span class="va">top5</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 35 x 7</span></span>
<span class="co">#&gt; <span style="color: #949494;"># Groups:   cluster [7]</span></span>
<span class="co">#&gt;       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene   </span>
<span class="co">#&gt;       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 8.39<span style="color: #949494;">e</span><span style="color: #BB0000;">-95</span>      0.672 1     0.866  1.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-90</span> 0       ENC1   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 2.99<span style="color: #949494;">e</span><span style="color: #BB0000;">-90</span>      0.751 0.928 0.529  4.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-86</span> 0       HOPX   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-67</span>      0.574 0.444 0.13   2.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-63</span> 0       CARTPT </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-64</span>      0.656 0.828 0.438  2.53<span style="color: #949494;">e</span><span style="color: #BB0000;">-60</span> 0       HPCAL1 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 5.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-62</span>      0.583 0.46  0.156  7.67<span style="color: #949494;">e</span><span style="color: #BB0000;">-58</span> 0       SAA1   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 7.59<span style="color: #949494;">e</span><span style="color: #BB0000;">-94</span>      0.554 1     0.971  1.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-89</span> 1       SNAP25 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-78</span>      0.608 1     0.894  2.05<span style="color: #949494;">e</span><span style="color: #BB0000;">-74</span> 1       SYT1   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 9.28<span style="color: #949494;">e</span><span style="color: #BB0000;">-67</span>      0.778 0.862 0.567  1.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-62</span> 1       DIRAS2 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span>      0.578 0.545 0.274  2.16<span style="color: #949494;">e</span><span style="color: #BB0000;">-36</span> 1       TBR1   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 5.05<span style="color: #949494;">e</span><span style="color: #BB0000;">-32</span>      0.585 0.871 0.664  7.59<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span> 1       SCGB2A2</span>
<span class="co">#&gt; <span style="color: #949494;"># i 25 more rows</span></span></code></pre></div>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 22621)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                              </span>
<span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] dplyr_1.0.9        Matrix_1.4-0       sp_1.5-0           SeuratObject_4.1.0</span>
<span class="co">#&gt; [5] Seurat_4.1.1       PRECAST_1.6.3      ProFAST_1.4        gtools_3.9.2.2    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] rappdirs_0.3.3              scattermore_0.8            </span>
<span class="co">#&gt;   [3] ggthemes_4.2.4              GiRaF_1.0.1                </span>
<span class="co">#&gt;   [5] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">#&gt;   [7] ggplot2_3.4.1               bit64_4.0.5                </span>
<span class="co">#&gt;   [9] knitr_1.37                  irlba_2.3.5                </span>
<span class="co">#&gt;  [11] DelayedArray_0.20.0         data.table_1.14.2          </span>
<span class="co">#&gt;  [13] rpart_4.1.16                KEGGREST_1.34.0            </span>
<span class="co">#&gt;  [15] RCurl_1.98-1.6              generics_0.1.2             </span>
<span class="co">#&gt;  [17] BiocGenerics_0.40.0         org.Mm.eg.db_3.14.0        </span>
<span class="co">#&gt;  [19] ScaledMatrix_1.2.0          cowplot_1.1.1              </span>
<span class="co">#&gt;  [21] RSQLite_2.2.10              RANN_2.6.1                 </span>
<span class="co">#&gt;  [23] future_1.26.1               bit_4.0.4                  </span>
<span class="co">#&gt;  [25] spatstat.data_3.0-0         xml2_1.3.3                 </span>
<span class="co">#&gt;  [27] httpuv_1.6.5                SummarizedExperiment_1.24.0</span>
<span class="co">#&gt;  [29] assertthat_0.2.1            viridis_0.6.2              </span>
<span class="co">#&gt;  [31] xfun_0.29                   hms_1.1.1                  </span>
<span class="co">#&gt;  [33] jquerylib_0.1.4             evaluate_0.15              </span>
<span class="co">#&gt;  [35] promises_1.2.0.1            fansi_1.0.4                </span>
<span class="co">#&gt;  [37] progress_1.2.2              dbplyr_2.1.1               </span>
<span class="co">#&gt;  [39] igraph_1.3.5                DBI_1.1.2                  </span>
<span class="co">#&gt;  [41] htmlwidgets_1.5.4           spatstat.geom_2.4-0        </span>
<span class="co">#&gt;  [43] stats4_4.1.2                purrr_0.3.4                </span>
<span class="co">#&gt;  [45] ellipsis_0.3.2              RSpectra_0.16-1            </span>
<span class="co">#&gt;  [47] ggpubr_0.4.0                backports_1.4.1            </span>
<span class="co">#&gt;  [49] DR.SC_3.4                   insight_0.17.0             </span>
<span class="co">#&gt;  [51] biomaRt_2.50.3              deldir_1.0-6               </span>
<span class="co">#&gt;  [53] sparseMatrixStats_1.6.0     MatrixGenerics_1.6.0       </span>
<span class="co">#&gt;  [55] vctrs_0.6.1                 SingleCellExperiment_1.16.0</span>
<span class="co">#&gt;  [57] Biobase_2.54.0              ROCR_1.0-11                </span>
<span class="co">#&gt;  [59] abind_1.4-5                 cachem_1.0.6               </span>
<span class="co">#&gt;  [61] withr_2.5.0                 progressr_0.10.1           </span>
<span class="co">#&gt;  [63] sctransform_0.3.3           prettyunits_1.1.1          </span>
<span class="co">#&gt;  [65] mclust_5.4.10               goftest_1.2-3              </span>
<span class="co">#&gt;  [67] cluster_2.1.2               lazyeval_0.2.2             </span>
<span class="co">#&gt;  [69] crayon_1.5.1                labeling_0.4.2             </span>
<span class="co">#&gt;  [71] pkgconfig_2.0.3             GenomeInfoDb_1.30.1        </span>
<span class="co">#&gt;  [73] nlme_3.1-155                vipor_0.4.5                </span>
<span class="co">#&gt;  [75] nnet_7.3-17                 rlang_1.1.0                </span>
<span class="co">#&gt;  [77] globals_0.15.0              lifecycle_1.0.3            </span>
<span class="co">#&gt;  [79] miniUI_0.1.1.1              filelock_1.0.2             </span>
<span class="co">#&gt;  [81] BiocFileCache_2.2.1         rsvd_1.0.5                 </span>
<span class="co">#&gt;  [83] rprojroot_2.0.3             polyclip_1.10-0            </span>
<span class="co">#&gt;  [85] matrixStats_0.62.0          lmtest_0.9-40              </span>
<span class="co">#&gt;  [87] carData_3.0-5               zoo_1.8-10                 </span>
<span class="co">#&gt;  [89] beeswarm_0.4.0              ggridges_0.5.3             </span>
<span class="co">#&gt;  [91] png_0.1-7                   viridisLite_0.4.1          </span>
<span class="co">#&gt;  [93] bitops_1.0-7                KernSmooth_2.23-20         </span>
<span class="co">#&gt;  [95] Biostrings_2.62.0           blob_1.2.2                 </span>
<span class="co">#&gt;  [97] DelayedMatrixStats_1.16.0   stringr_1.4.0              </span>
<span class="co">#&gt;  [99] parallelly_1.32.0           spatstat.random_2.2-0      </span>
<span class="co">#&gt; [101] rstatix_0.7.0               S4Vectors_0.32.3           </span>
<span class="co">#&gt; [103] ggsignif_0.6.3              beachmat_2.10.0            </span>
<span class="co">#&gt; [105] scales_1.2.1                memoise_2.0.1              </span>
<span class="co">#&gt; [107] magrittr_2.0.3              plyr_1.8.7                 </span>
<span class="co">#&gt; [109] ica_1.0-2                   zlibbioc_1.40.0            </span>
<span class="co">#&gt; [111] compiler_4.1.2              RColorBrewer_1.1-3         </span>
<span class="co">#&gt; [113] fitdistrplus_1.1-8          cli_3.2.0                  </span>
<span class="co">#&gt; [115] XVector_0.34.0              listenv_0.8.0              </span>
<span class="co">#&gt; [117] patchwork_1.1.1             pbapply_1.5-0              </span>
<span class="co">#&gt; [119] MASS_7.3-55                 mgcv_1.8-39                </span>
<span class="co">#&gt; [121] tidyselect_1.1.2            stringi_1.7.6              </span>
<span class="co">#&gt; [123] textshaping_0.3.6           highr_0.9                  </span>
<span class="co">#&gt; [125] yaml_2.3.6                  BiocSingular_1.10.0        </span>
<span class="co">#&gt; [127] ggrepel_0.9.1               grid_4.1.2                 </span>
<span class="co">#&gt; [129] sass_0.4.1                  tools_4.1.2                </span>
<span class="co">#&gt; [131] future.apply_1.9.0          rstudioapi_0.13            </span>
<span class="co">#&gt; [133] gridExtra_2.3               farver_2.1.1               </span>
<span class="co">#&gt; [135] Rtsne_0.16                  digest_0.6.29              </span>
<span class="co">#&gt; [137] rgeos_0.5-9                 shiny_1.7.1                </span>
<span class="co">#&gt; [139] Rcpp_1.0.10                 GenomicRanges_1.46.1       </span>
<span class="co">#&gt; [141] car_3.0-12                  broom_0.7.12               </span>
<span class="co">#&gt; [143] scuttle_1.4.0               performance_0.8.0          </span>
<span class="co">#&gt; [145] later_1.3.0                 harmony_0.1.0              </span>
<span class="co">#&gt; [147] RcppAnnoy_0.0.19            org.Hs.eg.db_3.14.0        </span>
<span class="co">#&gt; [149] httr_1.4.3                  AnnotationDbi_1.56.2       </span>
<span class="co">#&gt; [151] colorspace_2.1-0            XML_3.99-0.9               </span>
<span class="co">#&gt; [153] fs_1.5.2                    tensor_1.5                 </span>
<span class="co">#&gt; [155] reticulate_1.25             IRanges_2.28.0             </span>
<span class="co">#&gt; [157] splines_4.1.2               uwot_0.1.11                </span>
<span class="co">#&gt; [159] spatstat.utils_3.0-1        pkgdown_2.0.6              </span>
<span class="co">#&gt; [161] scater_1.25.1               plotly_4.10.0              </span>
<span class="co">#&gt; [163] systemfonts_1.0.4           xtable_1.8-4               </span>
<span class="co">#&gt; [165] jsonlite_1.8.0              R6_2.5.1                   </span>
<span class="co">#&gt; [167] pillar_1.9.0                htmltools_0.5.2            </span>
<span class="co">#&gt; [169] mime_0.12                   tictoc_1.2                 </span>
<span class="co">#&gt; [171] glue_1.6.2                  fastmap_1.1.0              </span>
<span class="co">#&gt; [173] BiocParallel_1.28.3         BiocNeighbors_1.12.0       </span>
<span class="co">#&gt; [175] codetools_0.2-18            furrr_0.3.1                </span>
<span class="co">#&gt; [177] utf8_1.2.3                  lattice_0.20-45            </span>
<span class="co">#&gt; [179] bslib_0.3.1                 spatstat.sparse_2.1-1      </span>
<span class="co">#&gt; [181] tibble_3.2.1                curl_4.3.2                 </span>
<span class="co">#&gt; [183] ggbeeswarm_0.6.0            leiden_0.4.2               </span>
<span class="co">#&gt; [185] limma_3.50.1                survival_3.2-13            </span>
<span class="co">#&gt; [187] CompQuadForm_1.4.3          rmarkdown_2.11             </span>
<span class="co">#&gt; [189] desc_1.4.0                  munsell_0.5.0              </span>
<span class="co">#&gt; [191] GenomeInfoDbData_1.2.7      reshape2_1.4.4             </span>
<span class="co">#&gt; [193] gtable_0.3.3                spatstat.core_2.4-4</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
