<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CoFAST: PBMC scRNA-seq data coembedding • ProFAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CoFAST: PBMC scRNA-seq data coembedding">
<meta property="og:description" content="ProFAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ProFAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CosMx.html">CoFAST: NSCLC CosMx data coembedding</a>
    </li>
    <li>
      <a href="../articles/FASTdlpfc.html">FAST: single DLPFC section</a>
    </li>
    <li>
      <a href="../articles/FASTdlpfc2.html">FAST: two DLPFC sections</a>
    </li>
    <li>
      <a href="../articles/FASTsimu.html">FAST: simulation</a>
    </li>
    <li>
      <a href="../articles/pbmc3k.html">CoFAST: PBMC scRNA-seq data coembedding</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/ProFAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CoFAST: PBMC scRNA-seq data coembedding</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2024-03-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/ProFAST/blob/HEAD/vignettes/pbmc3k.Rmd" class="external-link"><code>vignettes/pbmc3k.Rmd</code></a></small>
      <div class="hidden name"><code>pbmc3k.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the CoFAST workflow for the analysis of PBMC3k single-cell RNA sequencing dataset. In this vignette, the workflow of CoFAST consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Coembedding dimension reduction</li>
<li>Downstream analysis (i.e. , signature gene analysis, visualization of cell types and coembeddings)</li>
</ul>
<div class="section level2">
<h2 id="load-and-view-data">Load and view data<a class="anchor" aria-label="anchor" href="#load-and-view-data"></a>
</h2>
<p>We demonstrate the use of CoFAST to PBMC3k data that are in the SeuratData package, which can be downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span> <span class="co"># set a random seed for reproducibility.</span>
<span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="co">## filter the seurat_annotation is NA</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span><span class="op">)</span>
<span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">[</span>,<span class="va">idx</span><span class="op">]</span>
<span class="va">pbmc3k</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 13714 features across 2638 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (13714 features, 0 variable features)</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/ProFAST" class="external-link">ProFAST</a></span><span class="op">)</span> <span class="co"># load the package of FAST method</span>
<span class="co">#&gt; Loading required package: gtools</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'Seurat' was built under R version 4.1.3</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; Attaching sp</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>First, we normalize the data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></code></pre></div>
<p>Then, we select the variable genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="coembedding-using-non-centered-factor-model">Coembedding using non-centered factor model<a class="anchor" aria-label="anchor" href="#coembedding-using-non-centered-factor-model"></a>
</h2>
<p>We introduce how to use the non-centered factor model (NCFM) to perform coembedding for this scRNA-seq data. First, we determine the dimension of coembeddings. Here, we use the parallel analysis method to select the dimension.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diagnostic.cor.eigs.html">diagnostic.cor.eigs</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">q_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">dat_cor</span>, <span class="st">"q_est"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"q_est = "</span>, <span class="va">q_est</span>, <span class="st">'\n'</span><span class="op">)</span>
<span class="co">#&gt; q_est =  37</span></code></pre></div>
<p>Subsequently, we calculate coembeddings by utilizing NCFM, and observe that the <code>reductions</code> field acquires an additional component named <code>ncfm</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NCFM.html">NCFM</a></span><span class="op">(</span><span class="va">pbmc3k</span>, q <span class="op">=</span> <span class="va">q_est</span><span class="op">)</span>
<span class="va">pbmc3k</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 13714 features across 2638 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (13714 features, 2000 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: ncfm</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<p>In the following, we show how to find the signature genes based on comebeddings. First, we calculate the distance matrix.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pdistance.html">pdistance</a></span><span class="op">(</span><span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"ncfm"</span><span class="op">)</span>
<span class="co">#&gt; Calculate co-embedding distance...</span></code></pre></div>
<p>Next, we find the signature genes for each cell type</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Naive CD4 T Memory CD4 T   CD14+ Mono            B        CD8 T FCGR3A+ Mono </span>
<span class="co">#&gt;          697          483          480          344          271          162 </span>
<span class="co">#&gt;           NK           DC     Platelet </span>
<span class="co">#&gt;          155           32           14</span>
<span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span>
<span class="va">df_sig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find.signature.genes.html">find.signature.genes</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">df_sig_list</span><span class="op">)</span>
<span class="co">#&gt; List of 9</span>
<span class="co">#&gt;  $ B           :'data.frame':    1674 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:1674] 0.103 0.103 0.103 0.103 0.103 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:1674] 0.488 0.936 0.564 0.311 0.297 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:1674] "B" "B" "B" "B" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:1674] "VPREB3" "CD79A" "LINC00926" "TSPAN13" ...</span>
<span class="co">#&gt;  $ CD14+ Mono  :'data.frame':    2039 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:2039] 0.109 0.109 0.109 0.109 0.109 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:2039] 0.371 0.908 0.667 0.685 0.223 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:2039] "CD14+ Mono" "CD14+ Mono" "CD14+ Mono" "CD14+ Mono" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:2039] "ASGR1" "LGALS2" "CD14" "MS4A6A" ...</span>
<span class="co">#&gt;  $ CD8 T       :'data.frame':    1986 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:1986] 0.12 0.121 0.122 0.122 0.122 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:1986] 0.192 0.985 0.506 0.587 0.491 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:1986] "CD8 T" "CD8 T" "CD8 T" "CD8 T" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:1986] "LAG3" "CCL5" "CD8A" "GZMK" ...</span>
<span class="co">#&gt;  $ DC          :'data.frame':    3893 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:3893] 0.0939 0.0958 0.0963 0.0969 0.0988 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:3893] 0.812 0.281 0.125 0.5 0.438 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:3893] "DC" "DC" "DC" "DC" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:3893] "FCER1A" "CLIC2" "FLT3" "SERPINF1" ...</span>
<span class="co">#&gt;  $ FCGR3A+ Mono:'data.frame':    3010 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:3010] 0.0946 0.0946 0.0956 0.0962 0.0962 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:3010] 0.37 0.506 0.16 0.383 0.105 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:3010] "FCGR3A+ Mono" "FCGR3A+ Mono" "FCGR3A+ Mono" "FCGR3A+ Mono" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:3010] "CKB" "CDKN1C" "ICAM4" "MS4A4A" ...</span>
<span class="co">#&gt;  $ Memory CD4 T:'data.frame':    2308 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:2308] 0.104 0.104 0.104 0.104 0.105 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:2308] 0.213 0.422 0.263 0.172 0.195 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:2308] "Memory CD4 T" "Memory CD4 T" "Memory CD4 T" "Memory CD4 T" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:2308] "TNFRSF4" "AQP3" "CD40LG" "CRIP2" ...</span>
<span class="co">#&gt;  $ Naive CD4 T :'data.frame':    1703 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:1703] 0.103 0.103 0.103 0.103 0.103 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:1703] 0.436 0.197 0.152 0.336 0.156 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:1703] "Naive CD4 T" "Naive CD4 T" "Naive CD4 T" "Naive CD4 T" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:1703] "CCR7" "FHIT" "LINC00176" "LEF1" ...</span>
<span class="co">#&gt;  $ NK          :'data.frame':    2306 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:2306] 0.122 0.122 0.122 0.122 0.123 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:2306] 0.11 0.471 0.29 0.271 0.206 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:2306] "NK" "NK" "NK" "NK" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:2306] "KIR2DL3" "AKR1C3" "SH2D1B" "KLRF1" ...</span>
<span class="co">#&gt;  $ Platelet    :'data.frame':    1581 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ distance : num [1:1581] 0.114 0.115 0.115 0.115 0.116 ...</span>
<span class="co">#&gt;   ..$ expr.prop: num [1:1581] 0.571 0.857 0.929 0.857 0.429 ...</span>
<span class="co">#&gt;   ..$ label    : chr [1:1581] "Platelet" "Platelet" "Platelet" "Platelet" ...</span>
<span class="co">#&gt;   ..$ gene     : chr [1:1581] "LY6G6F" "AP001189.4" "GP9" "ITGA2B" ...</span></code></pre></div>
<p>Then, we obtain the top five signature genes and organize them into a data.frame. The colname <code>distance</code> means the distance between gene (i.e., VPREB3) and cells with the specific cell type (i.e., B cell), which is calculated based on the coembedding of genes and cells in the coembedding space. The distance is smaller, the association between gene and the cell type is stronger. The colname <code>expr.prop</code> represents the expression proportion of the gene (i.e., VPREB3) within the cell type (i.e., B cell). The colname <code>label</code> means the cell types and colname <code>gene</code> denotes the gene name. By the data.frame object, we know <code>VPREB3</code> is the one of the top signature gene of B cell.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get.top.signature.dat.html">get.top.signature.dat</a></span><span class="op">(</span><span class="va">df_sig_list</span>, ntop <span class="op">=</span> <span class="fl">5</span>, expr.prop.cutoff <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>
<span class="co">#&gt;            distance expr.prop      label      gene</span>
<span class="co">#&gt; VPREB3    0.1025561 0.4883721          B    VPREB3</span>
<span class="co">#&gt; CD79A     0.1026823 0.9360465          B     CD79A</span>
<span class="co">#&gt; LINC00926 0.1028375 0.5639535          B LINC00926</span>
<span class="co">#&gt; TSPAN13   0.1029338 0.3110465          B   TSPAN13</span>
<span class="co">#&gt; FCRLA     0.1029354 0.2965116          B     FCRLA</span>
<span class="co">#&gt; ASGR1     0.1093069 0.3708333 CD14+ Mono     ASGR1</span></code></pre></div>
<p>Next, we calculate the UMAP projections of coembeddings of cells and the selected signature genes.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembedding_umap.html">coembedding_umap</a></span><span class="op">(</span>
  <span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"ncfm"</span>, reduction.name <span class="op">=</span> <span class="st">"UMAP"</span>,
  gene.set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Furthermore, we visualize the cells and top five signature genes of B cell in the UMAP space of coembedding. We observe that the UMAP projections of the five signature genes are near to B cells, which indicates these genes are enriched in B cells.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## choose beutifual colors</span>
<span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="fu">PRECAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Light 13"</span>, n_colors <span class="op">=</span> <span class="fl">9</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembed_plot.html">coembed_plot</a></span><span class="op">(</span>
   <span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span>,
   gene_txtdata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">label</span><span class="op">==</span><span class="st">'B'</span><span class="op">)</span>, 
   cols<span class="op">=</span><span class="va">cols_cluster</span>,pt_text_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-12-2.png" width="768"></p>
<p>Then, we visualize the cells and top five signature genes of all involved cell types in the UMAP space of coembedding. We observe that the UMAP projections of the five signature genes are near to the corresponding cell type, which indicates these genes are enriched in the corresponding cells.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembed_plot.html">coembed_plot</a></span><span class="op">(</span>
   <span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span>,
   gene_txtdata <span class="op">=</span> <span class="va">dat</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>,
   pt_text_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">p2</span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<p>In addtion, we can fully take advantages of the visualization functions in <code>Seurat</code> package for visualization. The following is an example that visualizes the cell types on the UMAP space.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_type</span> <span class="op">&lt;-</span> <span class="va">cols_cluster</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols_type</span><span class="op">)</span><span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">'UMAP'</span>, cols<span class="op">=</span><span class="va">cols_type</span><span class="op">)</span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>Then, there is another example that we plot the two signature genes of B cell on UMAP space, in which we observed the high expression in B cells in constrast to other cell types.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">'UMAP'</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD79A"</span>, <span class="st">"VPREB3"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 22621)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                              </span>
<span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] purrr_0.3.4        sp_1.5-0           SeuratObject_4.1.0 Seurat_4.1.1      </span>
<span class="co">#&gt; [5] ProFAST_1.4        gtools_3.9.2.2    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] rappdirs_0.3.3              scattermore_0.8            </span>
<span class="co">#&gt;   [3] ggthemes_4.2.4              GiRaF_1.0.1                </span>
<span class="co">#&gt;   [5] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">#&gt;   [7] ggplot2_3.4.1               bit64_4.0.5                </span>
<span class="co">#&gt;   [9] knitr_1.37                  irlba_2.3.5                </span>
<span class="co">#&gt;  [11] DelayedArray_0.20.0         data.table_1.14.2          </span>
<span class="co">#&gt;  [13] rpart_4.1.16                KEGGREST_1.34.0            </span>
<span class="co">#&gt;  [15] RCurl_1.98-1.6              generics_0.1.2             </span>
<span class="co">#&gt;  [17] BiocGenerics_0.40.0         org.Mm.eg.db_3.14.0        </span>
<span class="co">#&gt;  [19] ScaledMatrix_1.2.0          cowplot_1.1.1              </span>
<span class="co">#&gt;  [21] RSQLite_2.2.10              RANN_2.6.1                 </span>
<span class="co">#&gt;  [23] future_1.26.1               bit_4.0.4                  </span>
<span class="co">#&gt;  [25] spatstat.data_3.0-0         xml2_1.3.3                 </span>
<span class="co">#&gt;  [27] httpuv_1.6.5                SummarizedExperiment_1.24.0</span>
<span class="co">#&gt;  [29] assertthat_0.2.1            viridis_0.6.2              </span>
<span class="co">#&gt;  [31] xfun_0.29                   hms_1.1.1                  </span>
<span class="co">#&gt;  [33] jquerylib_0.1.4             evaluate_0.15              </span>
<span class="co">#&gt;  [35] promises_1.2.0.1            fansi_1.0.4                </span>
<span class="co">#&gt;  [37] progress_1.2.2              dbplyr_2.1.1               </span>
<span class="co">#&gt;  [39] igraph_1.3.5                DBI_1.1.2                  </span>
<span class="co">#&gt;  [41] htmlwidgets_1.5.4           spatstat.geom_2.4-0        </span>
<span class="co">#&gt;  [43] stats4_4.1.2                ellipsis_0.3.2             </span>
<span class="co">#&gt;  [45] dplyr_1.0.9                 ggpubr_0.4.0               </span>
<span class="co">#&gt;  [47] backports_1.4.1             DR.SC_3.4                  </span>
<span class="co">#&gt;  [49] insight_0.17.0              biomaRt_2.50.3             </span>
<span class="co">#&gt;  [51] deldir_1.0-6                sparseMatrixStats_1.6.0    </span>
<span class="co">#&gt;  [53] MatrixGenerics_1.6.0        vctrs_0.6.1                </span>
<span class="co">#&gt;  [55] SingleCellExperiment_1.16.0 Biobase_2.54.0             </span>
<span class="co">#&gt;  [57] ROCR_1.0-11                 abind_1.4-5                </span>
<span class="co">#&gt;  [59] cachem_1.0.6                withr_2.5.0                </span>
<span class="co">#&gt;  [61] PRECAST_1.6.3               progressr_0.10.1           </span>
<span class="co">#&gt;  [63] sctransform_0.3.3           prettyunits_1.1.1          </span>
<span class="co">#&gt;  [65] mclust_5.4.10               goftest_1.2-3              </span>
<span class="co">#&gt;  [67] cluster_2.1.2               lazyeval_0.2.2             </span>
<span class="co">#&gt;  [69] crayon_1.5.1                labeling_0.4.2             </span>
<span class="co">#&gt;  [71] pkgconfig_2.0.3             GenomeInfoDb_1.30.1        </span>
<span class="co">#&gt;  [73] nlme_3.1-155                vipor_0.4.5                </span>
<span class="co">#&gt;  [75] nnet_7.3-17                 rlang_1.1.0                </span>
<span class="co">#&gt;  [77] globals_0.15.0              lifecycle_1.0.3            </span>
<span class="co">#&gt;  [79] miniUI_0.1.1.1              filelock_1.0.2             </span>
<span class="co">#&gt;  [81] BiocFileCache_2.2.1         pbmc3k.SeuratData_3.1.4    </span>
<span class="co">#&gt;  [83] SeuratData_0.2.2            rsvd_1.0.5                 </span>
<span class="co">#&gt;  [85] rprojroot_2.0.3             polyclip_1.10-0            </span>
<span class="co">#&gt;  [87] matrixStats_0.62.0          lmtest_0.9-40              </span>
<span class="co">#&gt;  [89] Matrix_1.4-0                carData_3.0-5              </span>
<span class="co">#&gt;  [91] zoo_1.8-10                  beeswarm_0.4.0             </span>
<span class="co">#&gt;  [93] ggridges_0.5.3              png_0.1-7                  </span>
<span class="co">#&gt;  [95] viridisLite_0.4.1           bitops_1.0-7               </span>
<span class="co">#&gt;  [97] KernSmooth_2.23-20          Biostrings_2.62.0          </span>
<span class="co">#&gt;  [99] blob_1.2.2                  DelayedMatrixStats_1.16.0  </span>
<span class="co">#&gt; [101] stringr_1.4.0               parallelly_1.32.0          </span>
<span class="co">#&gt; [103] spatstat.random_2.2-0       rstatix_0.7.0              </span>
<span class="co">#&gt; [105] S4Vectors_0.32.3            ggsignif_0.6.3             </span>
<span class="co">#&gt; [107] beachmat_2.10.0             scales_1.2.1               </span>
<span class="co">#&gt; [109] memoise_2.0.1               magrittr_2.0.3             </span>
<span class="co">#&gt; [111] plyr_1.8.7                  ica_1.0-2                  </span>
<span class="co">#&gt; [113] zlibbioc_1.40.0             compiler_4.1.2             </span>
<span class="co">#&gt; [115] RColorBrewer_1.1-3          fitdistrplus_1.1-8         </span>
<span class="co">#&gt; [117] cli_3.2.0                   XVector_0.34.0             </span>
<span class="co">#&gt; [119] listenv_0.8.0               patchwork_1.1.1            </span>
<span class="co">#&gt; [121] pbapply_1.5-0               MASS_7.3-55                </span>
<span class="co">#&gt; [123] mgcv_1.8-39                 tidyselect_1.1.2           </span>
<span class="co">#&gt; [125] stxBrain.SeuratData_0.1.1   stringi_1.7.6              </span>
<span class="co">#&gt; [127] textshaping_0.3.6           highr_0.9                  </span>
<span class="co">#&gt; [129] yaml_2.3.6                  BiocSingular_1.10.0        </span>
<span class="co">#&gt; [131] ggrepel_0.9.1               grid_4.1.2                 </span>
<span class="co">#&gt; [133] sass_0.4.1                  tools_4.1.2                </span>
<span class="co">#&gt; [135] future.apply_1.9.0          parallel_4.1.2             </span>
<span class="co">#&gt; [137] rstudioapi_0.13             gridExtra_2.3              </span>
<span class="co">#&gt; [139] farver_2.1.1                Rtsne_0.16                 </span>
<span class="co">#&gt; [141] digest_0.6.29               rgeos_0.5-9                </span>
<span class="co">#&gt; [143] FNN_1.1.3.1                 shiny_1.7.1                </span>
<span class="co">#&gt; [145] Rcpp_1.0.10                 GenomicRanges_1.46.1       </span>
<span class="co">#&gt; [147] car_3.0-12                  broom_0.7.12               </span>
<span class="co">#&gt; [149] scuttle_1.4.0               performance_0.8.0          </span>
<span class="co">#&gt; [151] later_1.3.0                 harmony_0.1.0              </span>
<span class="co">#&gt; [153] RcppAnnoy_0.0.19            org.Hs.eg.db_3.14.0        </span>
<span class="co">#&gt; [155] httr_1.4.3                  AnnotationDbi_1.56.2       </span>
<span class="co">#&gt; [157] colorspace_2.1-0            XML_3.99-0.9               </span>
<span class="co">#&gt; [159] fs_1.5.2                    tensor_1.5                 </span>
<span class="co">#&gt; [161] reticulate_1.25             IRanges_2.28.0             </span>
<span class="co">#&gt; [163] splines_4.1.2               uwot_0.1.11                </span>
<span class="co">#&gt; [165] spatstat.utils_3.0-1        pkgdown_2.0.6              </span>
<span class="co">#&gt; [167] scater_1.25.1               plotly_4.10.0              </span>
<span class="co">#&gt; [169] systemfonts_1.0.4           xtable_1.8-4               </span>
<span class="co">#&gt; [171] jsonlite_1.8.0              R6_2.5.1                   </span>
<span class="co">#&gt; [173] pillar_1.9.0                htmltools_0.5.2            </span>
<span class="co">#&gt; [175] mime_0.12                   tictoc_1.2                 </span>
<span class="co">#&gt; [177] glue_1.6.2                  fastmap_1.1.0              </span>
<span class="co">#&gt; [179] BiocParallel_1.28.3         BiocNeighbors_1.12.0       </span>
<span class="co">#&gt; [181] codetools_0.2-18            furrr_0.3.1                </span>
<span class="co">#&gt; [183] utf8_1.2.3                  lattice_0.20-45            </span>
<span class="co">#&gt; [185] bslib_0.3.1                 spatstat.sparse_2.1-1      </span>
<span class="co">#&gt; [187] tibble_3.2.1                curl_4.3.2                 </span>
<span class="co">#&gt; [189] ggbeeswarm_0.6.0            leiden_0.4.2               </span>
<span class="co">#&gt; [191] survival_3.2-13             CompQuadForm_1.4.3         </span>
<span class="co">#&gt; [193] rmarkdown_2.11              desc_1.4.0                 </span>
<span class="co">#&gt; [195] munsell_0.5.0               GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt; [197] reshape2_1.4.4              gtable_0.3.3               </span>
<span class="co">#&gt; [199] spatstat.core_2.4-4</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
